### PR TITLE
Updated releases notes for 1.19

### PR DESCRIPTION
**Pull request by XAMPPRocky**
This is waiting on any beta backports in the coming weeks, and the lib stabilisations pr.

[Rendered](https://github.com/Aaronepower/rust/blob/patch-2/RELEASES.md).

### CODE REVIEW

**Comment by Aaronepower** [comment]
In this release there has been a lot of error codes added by @GuillaumeGomez I was wondering should they each get a line item, or is the current fine?

File: RELEASES.md [hunk 1]
```diff
@@ -1,3 +1,144 @@
+Version 1.19.0 (2017-07-20)
+===========================
+
+Language
+--------
+
+- [Numeric fields can now be used for creating tuple structs.][36868]
+  For example `struct Point(u32, u32); let x = Foo { 0: 7, 1: 0 };`.
```
Comment (jtdowney) [inline RELEASES.md]: Wouldn't this be `let x = Point { 0: 7, 1: 0 };`

File: RELEASES.md [hunk 2]
```diff
@@ -1,3 +1,144 @@
...
+  the crate is being compiled in.
+  Example: `println!("cargo:rustc-env=FOO=bar");`][cargo/3929]
+- [Subcommands now replace the current process rather than spawning a new
+  child process][cargo/3970]
+- [Workspace members can now accept glob file patterns][cargo/3979]
+- [Added `--all` flag to the `cargo bench` subcommand to run benchmarks of all
+  the members in a given workspace.][cargo/3988]
+- [Updated `libssh2-sys` to 0.2.6][cargo/4008]
+- [Target directory path is now in the cargo metadata][cargo/4022]
+- [Cargo no longer checks the crates.io index locally][cargo/4026] This should
```
Comment (Nemo157) [inline RELEASES.md]: I would word this something like "Cargo no longer checks out a local working directory for the crates.io index"

File: RELEASES.md [hunk 3]
```diff
@@ -1,3 +1,144 @@
...
+Compatibility Notes
+-------------------
+
+- [`MutexGuard<T>` may only be `Sync` if `T` is `Sync`.][41624]
+- [`-Z` flags are now no longer allowed to be used on the stable
+  compiler.][41751] This has been a warning for a year previous to this.
+- [Ending a float literal with `._` is now a hard error.
+  Example: `42._` .][41946]
+- [Ending a str literal with an underscore is now a hard error
+  Example: `"foo"_`][41990]
```
Comment (ollie27) [inline RELEASES.md]: This didn't make it into 1.19.

File: RELEASES.md [hunk 4]
```diff
@@ -1,3 +1,144 @@
...
+- [Lifetimes named `'_` are no longer allowed.][36892] This was previously
+  a warning.
+
+Misc
+----
+
+- [Added `rust-windbg.cmd`][39983] for loading rust `.natvis` files in the
+  Windows Debugger.
+- [Rust Language Server is now packaged as a non default component with the
+  `.exe`, `.msi`, and `.pkg` installers][42306]
```
Comment (ollie27) [inline RELEASES.md]: I think this has been reverted: #42472

**Review by ollie27 — CHANGES_REQUESTED** [review]
Looks like a few things are missing:
* `From<Box<str>> for Box<[u8]>`  #41258
* `eprint!` and `eprintln!` #41192
* `closure_to_fn_coercion` #42162

File: RELEASES.md [hunk 5]
```diff
@@ -1,3 +1,144 @@
+Version 1.19.0 (2017-07-20)
+===========================
+
+Language
+--------
+
+- [Numeric fields can now be used for creating tuple structs.][36868]
```
Comment (petrochenkov) [inline RELEASES.md]: The correct PR is https://github.com/rust-lang/rust/pull/41145

File: RELEASES.md (hunk 3 above)
Comment (petrochenkov) [inline RELEASES.md]: This is also a warning, not error.

File: RELEASES.md [hunk 6]
```diff
@@ -1,3 +1,144 @@
...
+
+- [`MutexGuard<T>` may only be `Sync` if `T` is `Sync`.][41624]
+- [`-Z` flags are now no longer allowed to be used on the stable
+  compiler.][41751] This has been a warning for a year previous to this.
+- [Ending a float literal with `._` is now a hard error.
+  Example: `42._` .][41946]
+- [Ending a str literal with an underscore is now a hard error
+  Example: `"foo"_`][41990]
+- [Publicly exposing a private type is now a hard error][34537] This was
+  previously a warning.
```
Comment (petrochenkov) [inline RELEASES.md]: Nah, this is wrong. Most of https://github.com/rust-lang/rust/issues/34537 isn't going to be an error, only some tiny piece of it (public reexports of private enum variants) has become an error.

File: RELEASES.md [hunk 7]
```diff
@@ -1,3 +1,144 @@
...
+- [Type parameter defaults in trait impls and functions is now a
+  hard error][36887] This was previously a warning.
```
Comment (petrochenkov) [inline RELEASES.md]: This is still a warning, there was too much breakage (as you can see, the issue is still open).

File: RELEASES.md [hunk 8]
```diff
@@ -1,3 +1,144 @@
...
+- [`use` imports on a private `extern crate` in a module is now a hard
+  error.][36886] This was previously a warning.
```
Comment (petrochenkov) [inline RELEASES.md]: https://github.com/rust-lang/rust/issues/36886 is about any use of private extern crates outside of their module, not about imports or reexports.

File: RELEASES.md [hunk 9]
```diff
@@ -1,3 +1,144 @@
...
+
+- [Numeric fields can now be used for creating tuple structs.][36868]
+  For example `struct Point(u32, u32); let x = Point { 0: 7, 1: 0 };`.
+- [Macro recursion limit increased to 1024 from 64.][41676]
+- [Added lint for detecting unused macros.][41907]
+- [`loop` can now return a value with `break`.][42016]
+  For example: `let x = loop { break 7; };`
+- [C compatible `union`s are now available.][42068] They can only contain `Copy`
+  types and cannot have a `Drop` implementation.
+  Example: `union Foo { bar: u8 }`
```
Comment (mbrubeck) [inline RELEASES.md]: A union with two fields of different types might be a clearer example.

**Review by mbrubeck — COMMENTED** [review]
These look great!  Here are a few additions that I think would clarify some of them.

File: RELEASES.md [hunk 10]
```diff
@@ -1,3 +1,144 @@
...
+
+Compiler
+--------
+
+- [Added bootstrap support for Android.][41370]
```
Comment (mbrubeck) [inline RELEASES.md]: I suggest expanding this to "Add support for bootstrapping the Rust compiler toolchain on Android."

File: RELEASES.md [hunk 11]
```diff
@@ -1,3 +1,144 @@
...
+- [Change `arm-linux-androideabi` to correspond to the `armeabi`
+  official ABI.][41656]
```
Comment (mbrubeck) [inline RELEASES.md]: It might be useful to note that developers must now use `--target armv7-linux-androideabi` if they want to continue targeting `armeabi-v7a` ABI.

**Comment by cuviper** [comment]
Should this now mention LLVM4 support?  (i.e. rust-llvm is upgraded, and external LLVM4 works too.)

File: RELEASES.md [hunk 12]
```diff
@@ -1,3 +1,144 @@
...
+- [Macro recursion limit increased to 1024 from 64.][41676]
+- [Added lint for detecting unused macros.][41907]
+- [`loop` can now return a value with `break`.][42016]
+  For example: `let x = loop { break 7; };`
+- [C compatible `union`s are now available.][42068] They can only contain `Copy`
+  types and cannot have a `Drop` implementation.
+  Example: `union Foo { bar: u8 }`
+
+Compiler
+--------
```
Comment (TimNN) [inline RELEASES.md]: This section should probably mention the LLVM upgrade #40123.

**Comment by Aaronepower** [comment]
@cuviper @TimNN Is there anything to be said about using LLVM 4.0, or is `Now uses LLVM 4.0` enough?

**Comment by ranma42** [comment]
@brson on the rust side, XZ compression of the tarballs should ready. Should it be mentioned in the release notes or is it better to wait for the next rustup release?

**Comment by cuviper** [comment]
@Aaronepower There are two sides to it, both the version of LLVM that Rust is shipped with and the external versions it supports in custom builds.  For the [LLVM 3.9 update](https://github.com/rust-lang/rust/blob/ae3d3878ef831247ebb8babcd5efa345fdbb17eb/RELEASES.md#misc-6) it simply said, "The compiler can now be built against LLVM 3.9".

**Comment by brson** [comment]
Thanks a bunch @Aaronepower.

Added a link to the rendered notes to the OP.

@Aaronepower do any of these have RFC's that can be referenced for further info? I bet numeric fields, break return, unions, maybe closure coercions.

I see that 'compatibility notes' is not last this time. Any particular reason? I've put them last in the past to deemphasize them.

I don't see [custom CRT linkage](https://github.com/rust-lang/rust/issues/37406), [xz compression](https://github.com/rust-lang-nursery/rustup.rs/pull/1100).

I've added a milestone issue to make sure rustup with xz gets released: https://github.com/rust-lang/rust/issues/42542

**Comment by brson** [comment]
@ranma42 Yes, I think it should be mentioned. rustup will be out in time.

**Comment by brson** [comment]
@Aaronepower I think one line for the error codes is fine.

**Comment by alexcrichton** [comment]
ping @Aaronepower, just wanted to keep this on your radar!

**Comment by kennytm** [comment]
I think rust-lang/cargo#3954 (Add support for custom target-specific runners — `target.$triple.runner` in `.cargo/config`) is worth mentioning.

**Comment by Aaronepower** [comment]
@brson I've updated it with the RFC links, XZ compression, custom CRT linkage, and the *Compatibility Notes* section placement was not intentional and has been rearranged.

**Comment by brson** [comment]
There's a libs pr now: https://github.com/rust-lang/rust/pull/42745

**Comment by brson** [comment]
Here's a compatibility note for 1.19: https://github.com/rust-lang/rust/pull/42634

**Comment by brson** [comment]
Theres ARM fixes are probably noteworthy too https://github.com/rust-lang/rust/pull/42740

**Comment by Mark-Simulacrum** [comment]
Another compat note: https://github.com/rust-lang/rust/pull/41785

**Comment by Osspial** [comment]
The notes say that [`RangeArgument` and `Bound` are getting stabilized under the `ops` module ](https://github.com/Aaronepower/rust/blob/82f3a971fc2f7fe83b57466d90bdd0b4d6407630/RELEASES.md#stabilized-apis), but [testing them on the latest beta](https://is.gd/WLfnAg) gives a compile error. Is this an error on the release notes' part, or should they have been actually included in the beta?

**Comment by Aaronepower** [comment]
@Osspial The libs stabilisation pr which moves them to the `ops` module, hasn't been merged in yet.

**Comment by ishitatsuyuki** [comment]
The code was LLVM 4.0 compatible from at least one version before. The PR just fix some edge cases, tests and change the CI default.

**Comment by aidanhs** [comment]
@ishitatsuyuki that underplays the LLVM 4.0 PR a bit, but I agree that I'd reword it to something like "The compiler has upgraded to building against LLVM 4.0 by default"

**Comment by brson** [comment]
Another compat note: https://github.com/rust-lang/rust/issues/42460#issuecomment-313519570

**Comment by brson** [comment]
Might be worth mentioning that the `cargo-check` crate no longer works: https://github.com/rust-lang/rust/issues/42844#issuecomment-313557335

**Comment by brson** [comment]
If this one isn't fixed it will need a compat note: https://github.com/rust-lang/rust/issues/42635

That one I think is particularly alarming. It's basically breaking everything on Windows.

This is going to be a breaky release....

**Comment by brson** [comment]
LLVM is upgraded to 4.0.1 now https://github.com/rust-lang/rust/pull/42948

**Comment by brson** [comment]
It is possible the RangeArgument type will change names before release: https://github.com/rust-lang/rust/pull/43033

**Comment by brson** [comment]
RangeArgument is not in the release.

**Comment by brson** [comment]
Updated for all comments. https://github.com/rust-lang/rust/issues/42635 is slated to be fixed and doesn't need a compat note. We might wait to see that reverted and then this is probably good to go.

File: RELEASES.md [hunk 13]
```diff
@@ -1,3 +1,195 @@
...
+  path was relative regardless.
+- [Floating point constants in match patterns is now a hard error][36890]
+  This was previously a warning.
+- [Struct or enum constants that don't derive `PartialEq` & `Eq` used
+  match patterns is now a hard error][36891] This was previously a warning.
+- [Lifetimes named `'_` are no longer allowed.][36892] This was previously
+  a warning.
+- [From the pound escape, lines consisting of multiple `#`s are
+  now visible][41785]
+- [`for` loop desugaring has changed such that `break` does nat
```
Comment (mbrubeck) [inline RELEASES.md]: Typo: s/nat/not/

File: RELEASES.md [hunk 14]
```diff
@@ -1,3 +1,195 @@
...
+
+Misc
+----
+
+- [Added `rust-windbg.cmd`][39983] for loading rust `.natvis` files in the
+  Windows Debugger.
+- [Rust will now release XZ compressed packages][rust-installer/57]
+- [rustup will now prefer to download rust packages with
+  XZ compression][rustup/1100] over GZip packages.
+- [Added the ability to escape `#` in rust documentation][41785] By adding
```
Comment (mbrubeck) [inline RELEASES.md]: This change is mentioned twice (once here, then again in "Compatibility Notes" below).

Comment (brson) [inline RELEASES.md]: Yeah, I noticed that and decided that it's ok - it's a feature addition with compatibility concerns. *shrug*

**Comment by brson** [comment]
@steveklabnik since we're reverting this [change that affects nmake](https://github.com/rust-lang/rust/issues/42635), at least temporarily, to give more time for the ecosystem to adapt, it might be worth mentioning in the release notes so users are aware.

File: RELEASES.md [hunk 15]
```diff
@@ -1,3 +1,195 @@
...
+- [Floating point constants in match patterns is now a hard error][36890]
+  This was previously a warning.
+- [Struct or enum constants that don't derive `PartialEq` & `Eq` used
+  match patterns is now a hard error][36891] This was previously a warning.
+- [Lifetimes named `'_` are no longer allowed.][36892] This was previously
+  a warning.
+- [From the pound escape, lines consisting of multiple `#`s are
+  now visible][41785]
+- [`for` loop desugaring has changed such that `break` does not
+  affect type inference][42634]
```
Comment (Zoxc) [inline RELEASES.md]: This pull request is a fix for https://github.com/rust-lang/rust/pull/42265 which has never been stable.

This should say: `for` loop desugaring has changed such that  the iterator is not borrowed during the loop body.
#42265 should also be referenced.

It probably doesn't belong in compatibility notes since it should have no effect on Rust code.

Comment (brson) [inline RELEASES.md]: Thanks @Zoxc. Since it has no effect, I just removed it completely.

**Comment by brson** [comment]
Oh, if the jobserver patch is reverted, it will need to be removed from the relnotes.

**Comment by Aaronepower** [comment]
@brson I'm not seeing any revert PR, was this mentioned somewhere?

File: RELEASES.md [hunk 16]
```diff
@@ -1,3 +1,192 @@
...
+- [Change `arm-linux-androideabi` to correspond to the `armeabi`
+  official ABI.][41656] If you wish to continue targeting the `armeabi-v7a` ABI
+  you should use `--target armv7-linux-androideabi`.
+- [Fixed ICE when removing a source file between compilation sessions.][41873]
+- [Minor optimisation of string operations.][42037]
+- [Compiler error message is now `aborting due to previous error(s)` instead of
+  `aborting due to N previous errors`][42150] This was previously inaccurate and
+  would only count certain kinds of errors.
+- [The compiler now supports Visual Studio 2017][42225]
+- [The compiler is now be built against LLVM 4.0.1 by default][42948]
```
Comment (jimmycuadra) [inline RELEASES.md]: Typo: "is now be built"

**Comment by jimmycuadra** [comment]
In the compatibility notes, #34537 and #42460 are both linked with nearly identical descriptions. Do these represent the same underlying subject, and if so, should one of those links be removed?

**Comment by brson** [comment]
@Aaronepower jobserver was reverted [here](https://github.com/rust-lang/rust/pull/43208), which was actually merged [here](https://github.com/rust-lang/rust/pull/43114#issuecomment-315189458).

**Comment by brson** [comment]
@jimmycuadra yeah, they are both pretty much the same, probably we should only keep the link to the private_in_public PR.

File: RELEASES.md [hunk 17]
```diff
@@ -1,3 +1,187 @@
...
+
+- [`MutexGuard<T>` may only be `Sync` if `T` is `Sync`.][41624]
+- [`-Z` flags are now no longer allowed to be used on the stable
+  compiler.][41751] This has been a warning for a year previous to this.
+- [As a result of the `-Z` flag change, the `cargo-check` plugin no
+  longer works][42844]. Users should migrate to the built-in `check`
+  command, which has been available since 1.16.
+- [Ending a float literal with `._` is now a hard error.
+  Example: `42._` .][41946]
+- [Any use of a private `extern crate` outside of it's module is now a
```
Comment (sfackler) [inline RELEASES.md]: `its` not `it's` I think?

File: RELEASES.md [hunk 18]
```diff
@@ -1,3 +1,187 @@
...
+[42607]: https://github.com/rust-lang/rust/issues/42607
+[42740]: https://github.com/rust-lang/rust/pull/42740
+[42844]: https://github.com/rust-lang/rust/issues/42844
+[42948]: https://github.com/rust-lang/rust/pull/42948
+[RFC 1444]: https://github.com/rust-lang/rfcs/pull/1444
+[RFC 1506]: https://github.com/rust-lang/rfcs/pull/1506
+[RFC 1558]: https://github.com/rust-lang/rfcs/pull/1558
+[RFC 1624]: https://github.com/rust-lang/rfcs/pull/1624
+[RFC 1721]: https://github.com/rust-lang/rfcs/pull/1721
+[`Command::envs`]: https://doc.rust-lang.org/nightly/std/process/struct.Command.html#method.envs
```
Comment (steveklabnik) [inline RELEASES.md]: it'd be nice if this didn't link to nightly but it's not a blocker

**Comment by steveklabnik** [comment]
Github shows no conflict. Dunno whats up.

@bors: retry

maybe?

**Comment by Mark-Simulacrum** [comment]
This needs to be rebased, the conflict is presumably in a submodule.

**Comment by Aaronepower** [comment]
Closing in favour of a new PR without submodule problems. #43368
